File: LC2407_Longest_Increasing_Subsequence_2/longest_increasing_subsequence_2.py
from typing import List
class Solution:
    def lengthOfLIS(self, nums: List[int], k: int) -> int:
        n = len(nums)
        if n == 1:
            return 1
        
        dp_A = [1] * n # dp_A[i] = length of longest increasing subsequence in nums up to i if we do NOT include nums[i]
        dp_B = [1] * n # dp_B[i] = length of longest increasing subsequence in nums up to i if we include nums[i]
        dp_A[0] = 0
        dp_B[0] = 1
        lis = {nums[0]: 1} # key = nums[i], value = dp_B[i]

        for i in range(1, n):
            dp_A[i] = max(dp_A[i - 1], dp_B[i - 1])
            # To calculate dp_B[i], consider all the values from nums[i] - k to nums[i] - 1 (both inclusive)
            x = float('-inf')
            for j in range(k, 0, -1):
                if nums[i] - j in lis:
                    x = max(x, lis[nums[i] - j] + 1)
            dp_B[i] = max(1, x)
            lis[nums[i]] = dp_B[i]
        
        return max(dp_A[n - 1], dp_B[n - 1])

File: LC2407_Longest_Increasing_Subsequence_2/longest_increasing_subsequence_2.go
package main

import (
	"math"
)

func lengthOfLIS(nums []int, k int) int {
	n := len(nums)
	if n == 1 {
		return 1
	}
	dp_A := make([]int, n)
	dp_B := make([]int, n)
	dp_A[0] = 0
	dp_B[0] = 1
	lis := make(map[int]int)
	lis[nums[0]] = 1

	for i := 1; i < n; i++ {
		// Calculate dp_A[i]
		dp_A[i] = int(math.Max(float64(dp_A[i-1]), float64(dp_B[i-1])))

		// Calculate dp_B[i]
		// Consider all the values from nums[i] - k to nums[i] - 1 (both inclusive).
		x := math.Inf(-1)
		for j := k; j >= 1; j-- {
			val, exists := lis[nums[i]-j]
			if exists {
				x = math.Max(x, float64(val)+1.0)
			}
		}
		dp_B[i] = int(math.Max(1.0, x))
		lis[nums[i]] = dp_B[i]
	}

	return int(math.Max(float64(dp_A[n-1]), float64(dp_B[n-1])))
}
